### DISCUSSION TITLE
Define some metadata structure

### DISCUSSION THREAD

**Discussion by aborgna-q** [Quantinuum/hugr#268] [Specification]
Currently in the implementation the metadata for a node is just a `serde_yaml::Value`. This is really prone to cause collisions between (internal and external) users trying to store data there.

The spec doesn't define any structure for it, but it would be beneficial to at least require a HashMap<String, Value>.

Update: #661 Made the metadata always a HashMap, but the string keys do not have any structure yet.

**Comment by cqc-alec** [comment]
Agreed. Do we need to impose any more structure than `HashMap<String, Value>`?

**Reply by aborgna-q** [reply]
I'd say that's enough.

**Reply by ss2165** [reply]
I suggest rmp::Value rather than YAML, since that will be the most common source.
I agree with HashMap<String, Value> but we may wish to reserve some keys (or reserve a key _format_), and per reserved key we can impose some schema.

**Reply by aborgna-q** [reply]
This is currently a `serde_json::Map<String, serde_json::Value>`.

We could change it to use `rmp` instead.

**Reply by ss2165** [reply]
We should probably revisit the format (messagepack/json/whatever) after the wider data models discussion
